src/NUnitEngine/nunit.engine.core/RuntimeFramework.cs

```diff
@@ -85,84 +85,87 @@ public RuntimeFramework(RuntimeType runtime, Version version)
-                                break;
-                            case 4:
-                                this.ClrVersion = new Version(4, 0, 30319);
-                                break;
-                            default:
-                                ThrowInvalidFrameworkVersion(version);
-                                break;
-                        }
-                        break;
-                }
+            // All known framework versions have either two components or
+            // three. If three, then the Build is currently less than 3.
+            return v.Build < 3 && v.Revision == -1;
         }
 
-        private static void ThrowInvalidFrameworkVersion(Version version)
+        private Version GetClrVersionForFramework(Version frameworkVersion)
         {
-            throw new ArgumentException("Unknown framework version " + version.ToString(), "version");
+            switch (Runtime)
+            {
+                case RuntimeType.Net:
+                case RuntimeType.Any:
+                    switch (frameworkVersion.Major)
+                    {
+                        case 1:
+                            switch (frameworkVersion.Minor)
+                            {
+                                case 0:
+                                    return new Version(1, 0, 3705);
+                                case 1:
+                                    return new Version(1, 1, 4322);
+                            }
+                            break;
+                        case 2:
+                        case 3:
+                            return new Version(2, 0, 50727);
+                        case 4:
+                            return new Version(4, 0, 30319);
+                    }
+                    break;
+                case RuntimeType.Mono:
+                    switch (frameworkVersion.Major)
+                    {
+                        case 1:
+                            return new Version(1, 1, 4322);
+                        case 2:
+                        case 3:
+                            return new Version(2, 0, 50727);
+                        case 4:
+                            return new Version(4, 0, 30319);
+                    }
+                    break;
+            }
+
+            throw new ArgumentException("Unknown framework version " + frameworkVersion.ToString(), "version");
         }
 
-        private void InitFromClrVersion(Version version)
+        private Version GetFrameworkVersionForClr(Version clrVersion)
         {
-            this.FrameworkVersion = new Version(version.Major, version.Minor);
-            this.ClrVersion = version;
-            if (Runtime == RuntimeType.Mono && version.Major == 1)
-                this.FrameworkVersion = new Version(1, 0);
+            return Runtime == RuntimeType.Mono && clrVersion.Major == 1
+                ? new Version(1, 0)
+                : new Version(clrVersion.Major, clrVersion.Minor);
```

Comment:
Am I reading this right, that this is to detect Mono 1.0? This is the sort of complexity I'd love to strip out, in a future iteration! 😅

Reply:
in principle, the CLR version doesn't necessarily have obvious relation to the framework version. You could have SomeFramework 1.0 with a CLR of 5.1234.5678. Mono was merely the first place where this showed up - Mono 1.0 is essentially equivalent to MS .NET 1.1.

Easiest way to simplify IMO is to figure out how to do without the CLR version entirely. In the meantime, I'm thinking of a change in my own engine where the individual Runtimes figure out the CLR version, so e.g. Mono.Runtime would have a method that allowed deducing the CLR version from the framework version. Otherwise, this code will get a lot more complicated for .NET Core and other runtimes we add.